### PR TITLE
Random survival forest analytics view

### PR DESCRIPTION
**Pull request by hideaki**
# Description
- Random survival forest analytics view
- Permutation importance for survival forest and Cox regression
- Partial Dependence for survival forest and Cox regression

# Checklist
Make sure you have performed following items before submitting this pull request.
If not, please describe the reason.  

- [ ] Add test cases for this fix/enhancement
- [ ] Pass devtools::check()
- [ ] Pass devtools::test()
- [ ] Test installing from github
- [x] Tested with Exploratory

### CODE REVIEW

File: R/build_coxph.R [hunk 1]
```diff
@@ -62,6 +62,190 @@ build_coxph <- function(data, formula, max_categories = NULL, min_group_size = N
               ...)
 }
 
+partial_dependence.coxph_exploratory <- function(fit, time_col, vars = colnames(data),
```
Comment (hideaki) [inline R/build_coxph.R:65]: Add comment

File: R/build_coxph.R [hunk 2]
```diff
@@ -62,6 +62,190 @@ build_coxph <- function(data, formula, max_categories = NULL, min_group_size = N
...
+    for (l in 0:(num_event-1)) {
+      subtraction_term <- subtraction_term + log(sum_theta_risk - l/num_event*sum_theta_event)
+    }
+    contrib <- sum_lp_event - subtraction_term
+    contrib
+  }))
+  sum(purrr::flatten_dbl(tmp_df$contrib))
+}
+
+calc_permutation_importance_coxph <- function(fit, time_col, status_col, vars, data) {
```
Comment (hideaki) [inline R/build_coxph.R:236]: Add comment

File: R/build_coxph.R [hunk 3]
```diff
@@ -362,3 +657,42 @@ glance.coxph_exploratory <- function(x, pretty.name = FALSE, ...) { #TODO: add t
...
   ret
 }
+
+#' @export
+augment.coxph_exploratory <- function(x, ...) {
+  if ("error" %in% class(x)) {
+    ret <- data.frame(Note = x$message)
+    return(ret)
+  }
+  data <- x$source_data
```
Comment (hideaki) [inline R/build_coxph.R:667]: Add TODO comment to handle training/test split.

File: R/build_lm.R [hunk 4]
```diff
@@ -111,6 +111,25 @@ vif <- function(mod, ...) {
     result
 }
 
+calc_vif <- function(model) {
```
Comment (hideaki) [inline R/build_lm.R:114]: Add comment

File: R/survival_forest.R [hunk 5]
```diff
@@ -0,0 +1,486 @@
+calc_survival_curves_with_strata <- function(df, time_col, status_col, vars) {
```
Comment (hideaki) [inline R/survival_forest.R:1]: Add comment.

File: R/survival_forest.R [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+    ret <- broom:::tidy.survfit(fit)
+    ret
+  })
+  ret <- data.table::rbindlist(curve_dfs_list)
+  ret <- ret %>% tidyr::separate(strata, into = c('variable','value'), sep='=') %>% rename(survival=estimate, period=time)
+  ret <- ret %>% dplyr::mutate(chart_type = chart_type_map[variable])
+  ret
+}
+
+partial_dependence.ranger_survival_exploratory <- function(fit, time_col, vars = colnames(data),
```
Comment (hideaki) [inline R/survival_forest.R:27]: Add comment.

File: R/build_coxph.R [hunk 7]
```diff
@@ -62,6 +62,190 @@ build_coxph <- function(data, formula, max_categories = NULL, min_group_size = N
...
+    high <- as.data.frame(t(res %>% select(starts_with('conf.high.'))))
+    high <- high %>% dplyr::rename_at(vars(starts_with('V')), funs(stringr::str_replace(., 'V', 'H')))
+    low <- as.data.frame(t(res %>% select(starts_with('conf.low.'))))
+    low <- low %>% dplyr::rename_at(vars(starts_with('V')), funs(stringr::str_replace(., 'V', 'L')))
+    res <- dplyr::bind_cols(est, high, low)
+    res
+  }
+
+  aggregate.fun <- function(x) {
+    mean(x)
```
Comment (kei51e) [inline R/build_coxph.R:96]: Need NA handling?

File: R/build_coxph.R [hunk 8]
```diff
@@ -62,6 +62,190 @@ build_coxph <- function(data, formula, max_categories = NULL, min_group_size = N
...
+  tmp_df <- tmp_df %>% dplyr::group_by(time) %>%
+    dplyr::summarize(sum_lp_event = sum(lp*status),
+                     sum_theta_event = sum(theta*status),
+                     num_event = sum(status),
+                     sum_theta = sum(theta))
+
+  tmp_df <- tmp_df %>% dplyr::mutate(sum_theta_risk = rev(cumsum(rev(sum_theta))))
+
+  tmp_df <- tmp_df %>% dplyr::mutate(contrib = purrr::pmap(list(sum_lp_event, sum_theta_event, num_event, sum_theta_risk),
+                                                    function(sum_lp_event, sum_theta_event, num_event, sum_theta_risk) {
```
Comment (kei51e) [inline R/build_coxph.R:223]: Please add some description about what this anonymous function does.

File: R/build_coxph.R [hunk 9]
```diff
@@ -62,6 +62,190 @@ build_coxph <- function(data, formula, max_categories = NULL, min_group_size = N
...
+calc_efron_log_likelihood <- function(lp, time, status) {
+  if (is.data.frame(time)) { # Since mmpf::permutationImportance passes down time as tibble, convert it to vector. TODO: Do this at more appropriate place.
+    time <- time[[1]]
+  }
+  tmp_df <- tibble::tibble(time=time,
+                           status=as.numeric(status), # Handle logical status. TODO: Handle it better.
+                           lp=lp,
+                           theta=exp(lp))
+  tmp_df <- tmp_df %>% dplyr::group_by(time) %>%
+    dplyr::summarize(sum_lp_event = sum(lp*status),
```
Comment (kei51e) [inline R/build_coxph.R:215]: Make sure those values have no NA, or NA is one of expected values as a result of sum().

File: R/build_coxph.R [hunk 10]
```diff
@@ -269,6 +481,36 @@ build_coxph.fast <- function(df,
       rf$terms_mapping <- names(name_map)
       names(rf$terms_mapping) <- name_map
       rf$sampled_nrow <- sampled_nrow
+
+      rf$permutation_importance <- calc_permutation_importance_coxph(rf, clean_time_col, clean_status_col, c_cols, df)
+
+      # get importance to decice variables for partial dependence
```
Comment (kei51e) [inline R/build_coxph.R:487]:
```suggestion
      # get importance to decide variables for partial dependence
```

File: R/build_coxph.R [hunk 11]
```diff
@@ -62,6 +62,190 @@ build_coxph <- function(data, formula, max_categories = NULL, min_group_size = N
...
+
+# Efron's partial log likelihood approximation.
+# We use this as the cost function for permutation importance of Cox regression.
+# References:
+# http://www.math.ucsd.edu/~rxu/math284/slect5.pdf
+# https://bydmitry.github.io/efron-tensorflow.html
+# https://en.wikipedia.org/wiki/Proportional_hazards_model
+# 
+# lp - Linear predictor
+calc_efron_log_likelihood <- function(lp, time, status) {
```
Comment (kei51e) [inline R/build_coxph.R:206]: Add a test to make sure the function works as expected.

File: R/build_coxph.R [hunk 12]
```diff
@@ -269,6 +481,36 @@ build_coxph.fast <- function(df,
       rf$terms_mapping <- names(name_map)
```
Comment (kei51e) [inline R/build_coxph.R:481]: Add todo to rename 'rf' to something more meaningful (probably 'model').

File: R/build_coxph.R [hunk 13]
```diff
@@ -289,43 +531,96 @@ build_coxph.fast <- function(df,
...
+      # as.character is to be safe by converting from factor. With factor, reverse mapping result will be messed up.
+      ret$term <- x$terms_mapping[as.character(ret$term)]
+      ret
+    },
+    partial_dependence_survival_curve = {
+      ret <- x$partial_dependence
+      ret <- ret %>% dplyr::group_by(variable) %>% tidyr::nest() %>%
+        mutate(data = purrr::map(data,function(df){ # Show only 5 lines out of 9 lines for survival curve.
+          if (df$chart_type[[1]] == 'line') {
+            df %>% dplyr::mutate(value_index=as.integer(forcats::fct_inorder(value))) %>% dplyr::filter(value_index %% 2 == 1) %>% dplyr::mutate(value_index=ceiling(value_index/2))
```
Comment (kei51e) [inline R/build_coxph.R:553]: Add a comment why %%2.

File: R/survival_forest.R [hunk 14]
```diff
@@ -0,0 +1,486 @@
...
+  status_col <- dplyr::select_var(names(df), !! rlang::enquo(status))
+  # this evaluates select arguments like starts_with
+  selected_cols <- dplyr::select_vars(names(df), !!! rlang::quos(...))
+
+  grouped_cols <- grouped_by(df)
+
+  # remove grouped col or time/status col
+  selected_cols <- setdiff(selected_cols, c(grouped_cols, time_col, status_col))
+
+  if(test_rate < 0 | 1 < test_rate){
```
Comment (kei51e) [inline R/survival_forest.R:150]: Centralize the preprocessing logic.

File: R/survival_forest.R [hunk 15]
```diff
@@ -0,0 +1,486 @@
...
+      test_index <- sample_df_index(source_data, rate = test_rate, ordered = (test_split_type == "ordered"))
+      df <- safe_slice(source_data, test_index, remove = TRUE)
+      if (test_rate > 0) {
+        df_test <- safe_slice(source_data, test_index, remove = FALSE)
+        unknown_category_rows_index_vector <- get_unknown_category_rows_index_vector(df_test, df)
+        df_test <- df_test[!unknown_category_rows_index_vector, , drop = FALSE] # 2nd arg must be empty.
+        unknown_category_rows_index <- get_row_numbers_from_index_vector(unknown_category_rows_index_vector)
+      }
+
+      # build formula for lm
```
Comment (kei51e) [inline R/survival_forest.R:348]:
```suggestion
      # build formula for survival forest.
```

File: R/build_coxph.R (hunk 1 above)
Comment (hideaki) [reply R/build_coxph.R:65]: Done.

File: R/build_coxph.R (hunk 2 above)
Comment (hideaki) [reply R/build_coxph.R:236]: Done.

File: R/build_coxph.R (hunk 3 above)
Comment (hideaki) [reply R/build_coxph.R:667]: Done

File: R/build_lm.R (hunk 4 above)
Comment (hideaki) [reply R/build_lm.R:114]: Done.

File: R/survival_forest.R (hunk 5 above)
Comment (hideaki) [reply R/survival_forest.R:1]: Done.

File: R/survival_forest.R (hunk 6 above)
Comment (hideaki) [reply R/survival_forest.R:27]: Done.

File: R/build_coxph.R (hunk 7 above)
Comment (hideaki) [reply R/build_coxph.R:96]: Added it just in case.

File: R/build_coxph.R (hunk 8 above)
Comment (hideaki) [reply R/build_coxph.R:223]: Done

File: R/build_coxph.R (hunk 9 above)
Comment (hideaki) [reply R/build_coxph.R:215]: Added na.rm = TRUE just in case.

File: R/build_coxph.R (hunk 11 above)
Comment (hideaki) [reply R/build_coxph.R:206]: Added TODO comment.

File: R/build_coxph.R (hunk 12 above)
Comment (hideaki) [reply R/build_coxph.R:481]: Just replaced their names.

File: R/build_coxph.R (hunk 13 above)
Comment (hideaki) [reply R/build_coxph.R:553]: Done.

File: R/survival_forest.R (hunk 14 above)
Comment (hideaki) [reply R/survival_forest.R:150]: Added TODO comment.

**Pull request merged**